cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/WorkflowDataset.java

```diff
@@ -204,6 +207,78 @@ public int compare(WorkflowDataset.WorkflowRunRecord o1, WorkflowDataset.Workflo
     return programToRunRecord.values();
   }
 
+  @Nullable
+  public WorkflowRunRecord getRecord(Id.Workflow id, String runId) {
+    RunId pid = RunIds.fromString(runId);
+    long startTime = RunIds.getTime(pid, TimeUnit.SECONDS);
+    MDSKey mdsKey = new MDSKey.Builder().add(id.getNamespaceId())
+      .add(id.getApplicationId()).add(id.getId()).add(startTime).build();
+    byte[] startRowKey = mdsKey.getKey();
+    Scan scan = new Scan(startRowKey, null);
+
+    WorkflowRunRecord workflowRunRecord;
+    Scanner scanner = table.scan(scan);
+    Row indexRow = scanner.next();
+    if (indexRow == null) {
+      return null;
+    }
+    Map<byte[], byte[]> columns = indexRow.getColumns();
+    String workflowRunId = Bytes.toString(columns.get(RUNID));
+    long timeTaken = Bytes.toLong(columns.get(TIME_TAKEN));
+
+    List<ProgramRun> actionRunsList = GSON.fromJson(Bytes.toString(columns.get(NODES)), PROGRAM_RUNS_TYPE);
+    workflowRunRecord = new WorkflowRunRecord(workflowRunId, timeTaken, actionRunsList);
+
+    if (!runId.equals(workflowRunRecord.getWorkflowRunId())) {
+      return null;
+    } else {
+      return workflowRunRecord;
+    }
+  }
+
+  Map<String, WorkflowRunRecord> getDetailsOfRange(Id.Workflow workflow, String runId, int count, long timeInterval) {
+    Map<String, WorkflowRunRecord> mainRunRecords = getNeighbors(workflow, runId, count, timeInterval);
+    WorkflowRunRecord workflowRunRecord = getRecord(workflow, runId);
+    if (workflowRunRecord != null) {
+      mainRunRecords.put(workflowRunRecord.getWorkflowRunId(), workflowRunRecord);
+    }
+    return mainRunRecords;
+  }
+
+  private Map<String, WorkflowRunRecord> getNeighbors(Id.Workflow id, String runId, int count, long timeInterval) {
+    RunId pid = RunIds.fromString(runId);
+    long startTime = RunIds.getTime(pid, TimeUnit.SECONDS);
+    Map<String, WorkflowRunRecord> workflowRunRecords = new HashMap<>();
+    for (int i = (-1 * count); i <= count; i++) {
+      long prevStartTime = startTime + (i * timeInterval);
+      MDSKey mdsKey = new MDSKey.Builder().add(id.getNamespaceId())
+        .add(id.getApplicationId()).add(id.getId()).add(prevStartTime).build();
+      byte[] startRowKey = mdsKey.getKey();
+      Scan scan = new Scan(startRowKey, null);
+      Scanner scanner = table.scan(scan);
+      Row indexRow = scanner.next();
+      if (indexRow == null) {
+        return workflowRunRecords;
+      }
+      mdsKey = new MDSKey(indexRow.getRow());
+      MDSKey.Splitter splitter = mdsKey.split();
+      splitter.skipString();
+      splitter.skipString();
+      splitter.skipString();
```

Comment:
i see its private.. then you can do something like this
``
byte[] rowKey = indexRow.getRow();
// ByteBuffer.wrap(byte[], offset, length)
long time = ByteBuffer.wrap(rowKey, rowKey.length() - Bytes.SIZEOF_LONG, SIZEOF_LONG).getLong()
``
since we know the time is at the end and the length of it (long), doing it this way is robust.